stdlib/args.ysh

```diff
@@ -24,140 +24,159 @@
 }
 
-proc __arg (name, desc ; type ; default Any = null) {
-  # Type is always string?  Or support conversion?
-  # I guess there's no reason not to
-  #
-  # Int or Float
-
-  echo 'arg'
+proc flag (short, long; type='bool', default=null) {
+  ## Declare a flag within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :spec {
+  ##     flag -v --verbose
+  ##     flag -n --count ('int', default=1)
+  ##     flag -f --file ('str')
+  ##   }
+
+  # TODO: help, make type take in a type object (ie. Bool, Int, Str)
+  json write --pretty=F ({ tag: 'flag', node: { short, long, type, default, help: null } })
 }
 
-# TODO: create a builtin?
-func getProc(x) {
-  return (x)
+proc arg (name) {
+  ## Declare a positional argument within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :spec {
+  ##     arg file
+  ##   }
+
+  # TODO: type, default and help
+  json write --pretty=F ({ tag: 'arg', node: { name, type: 'str', default: null, help: null } })
 }
-# TODO: builtin?
-# Is this different than shvar?  How?
-# Is it a global or a local?
 
-# Maybe we can rename shvar IFS=: (foo={}) { }
-#
-# - shvar does LocalOnly
-# - but 'proc arg' will push a new stack frame?
-#   - so we need something like setref?
-#
-
-proc bind-vars (; ; ...bindings; block) {
-  echo
+proc rest (name) {
+  ## Take the remaining positional arguments within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :grep {
+  ##     arg query
+  ##     rest files
+  ##   }
+
+  json write --pretty=F ({ tag: 'rest', node: name })
 }
 
-hay define Documented
-
-Documented {  # Do we inherit the name of the proc?  Or $path/$procName?
-
-  # Do we want a longer way of documenting APIs?
-  # Document every argument?
-  # I like Go's docs
-  desc = '''
-  These are my docs
-  '''
-
-  proc p {
-    echo
+func __args_getFlagName(flag) {
+  if (flag.long) {
+    return (flag.long[2:])
+  } elif (flag.short) {
+    return (flag.short[1:])
   }
 
+  error ('No flag.long or flag.short')
 }
 
-proc Args(spec Ref ; ; ; block) {
-
-  var more_procs = {
-    flag: getProc('__flag'),
-    arg: getProc('__arg'),
-  }
-
-  var this = {}
+func parseArgs(spec, argv) {
+  ## Given a spec created by `Args`. Parse an array of strings `argv` per that
+  ## spec.
+  ##
+  ## See `Args` for examples of use.
 
-  # _this is part of this frame -- it's a duplicate that callers can look up I
-  # guess
-  # Well then all you need is getvar() ?  You don't even need bind-vars?
-  bind-vars (_this=this) {
-    # So now 'arg' and 'flag' can see _this?
+  var i = 0
+  var positionalPos = 0
+  var argc = len(argv)
+  var args = {}
+  var rest = []
+
+  # Would like to have: var value
+  var value = null
+  while (i < argc) {
+    var arg = argv[i]
+    if (arg->startswith('-')) {
+      for flag in (spec.flags) {
+        if ( (flag.short and flag.short === arg) or
+             (flag.long and flag.long === arg) ) {
+          case (flag.type) {
+            ('bool') | (null) { setvar value = true }
+            int {
```

Comment:
This could be `('int')` to be more consistent with the line above

I guess it is a bit subtle to have an unquoted word -- it can also be `'int'`